### PR TITLE
Refactor combineReduer's Unexpected State Shape Warning

### PR DESCRIPTION
**Pull request by jridgewell**
The shape of the previous state is irrelevant, what we’re actually
testing is that state has the same shape as our reducer object
`finalReducers`.

Also fixes the algorithm to be O(n) by using `#hasOwnProperty` vs doing
an array search O(n<sup>2</sup>).

### CODE REVIEW

**Comment by acdlite** [comment]
This looks good to me. :+1: for using `hasOwnProperty` especially.

**Comment by acdlite** [comment]
Someone else should review this too, though, since I'm not as familiar with this module.

File: src/utils/combineReducers.js [hunk 1]
```diff
@@ -15,8 +15,8 @@ function getUndefinedStateErrorMessage(key, action) {
   )
 }
 
-function getUnexpectedStateKeyWarningMessage(inputState, outputState, action) {
-  var reducerKeys = Object.keys(outputState)
+function getUnexpectedStateShapeWarningMessage(inputState, reducer, action) {
```
Comment (gaearon) [inline src/utils/combineReducers.js]: The argument should probably be called `reducers`, not `reducer`.

**Comment by gaearon** [comment]
I think @ellbee should remember this better than me.
Basically we need to make sure we don't regress on the warning we introduced in https://github.com/rackt/redux/issues/715#issuecomment-139843090 and https://github.com/rackt/redux/pull/720.

**Comment by ellbee** [comment]
I have just been playing with this, and I am pretty sure that this behaves the same as what is currently there.

**Comment by gaearon** [comment]
Leaving this for you to review then :+1:

File: src/utils/combineReducers.js (hunk 1 above)
Comment (jridgewell) [inline src/utils/combineReducers.js]: Updated.

**Comment by ellbee** [comment]
So, the non-cosmetic changes in this PR are the switch to `hasOwnProperty` and passing an empty object when initial state is not specified. I did some testing and they both look good. Thanks @jridgewell.

**Comment by gaearon** [comment]
Out in 3.0.6, thanks!

**Pull request merged**